docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+
+In most applications of the `Store`, it is preferable to have reducers perform work on the main thread. This helps to simplify the levels of async tasks that need to be synchronised with other async tasks that are already on the main thread.
+
+## Motivation
+
+Android embedders use the main thread for UI, user-facing, or touch handling work. For example, notifying UI components when IO from storage layers have completed, an engine's tasks that can happen on a separate thread, or global-level state updates for different components to observe.
+
+When components dispatch actions to the store, these actions are performed on an independant single thread dispatcher in the `Store` to avoid overloading the main thread with heavy work that might be performed during the `reduce` or in a `Middleware`. In practice, these actions have been short and fast so they do not cause overhead. Work done in a `Middleware` that can be slow, like I/O,  is put onto separate Dispatchers. The performance optimization to switch to a `Store` thread, requires that components which are always run on the main thread, to ensure synchronisation is now kept between the main thread and the store thread for observers of the `State`.
+
+There are some advantages to this change:
+
+* Simplicity for main-thread `Store`s.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads.
+
+## Guide-level explanation
+
+There are two possible ways to allow for this behaviour during the store creation:
+
+1. Pass in a `CoroutineDispatcher`.
+2. Configurable set of pre-defined options to choose.
+
+Passing in a `CoroutineDispatcher` has the flexibility to allow any preferred dispatcher to be used (e.g. `Dispatchers.IO` or Android Component's `Dispatchers.Cached`),  however that could lead to complications on choosing the wrong dispatcher. Ensuring that testing is also done with the same dispatcher configuration can also be difficult with this pattern.
+
+With a predefined set of options, we have a smaller surface area of changes that we want to maintain, possibly starting off with either the main dispatcher or the current single threaded dispatcher. We can continue to use our `StoreThreadFactory` and disable it when we switch to the `Main` dispatcher.
+
+Below are examples of what either surface would look like to a consumer of an "AppStore":
+
+```kotlin
+// Pass in a `CoroutineDispatcher`.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    dispatcher = Dispatchers.Main // choose a preferable dispatcher.
+)
+
+// Configurable set of pre-defined options to choose.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    runOnMain = true, // a boolean or enum value to signify the dispatcher used.
```

Comment:
Of these options, I think an enum value would be preferable here, but what benefit does this then provide over using the finite set already defined on `Dispatchers`? It seems as potentially error-prone to me.

That is to say, my preference would be to allow for `Dispatchers` directly.

Reply:
The boolean/naming is to show to an example of some kind of option, this is not the final name/version of it.

See the comment below about using any `Dispatcher`.

Reply:
> Is that then the goal of your proposal, and do we then still need the Dispatchers example as well?

Both examples were add for the sake of discussion to show that there will be some form of configuration added. It sounds like there is more preference towards the second option, so I can remove the one above if it is causing confusion.